nexus/networking/src/firewall_rules.rs

```diff
@@ -439,3 +486,100 @@ pub async fn plumb_service_firewall_rules(
     .await?;
     Ok(())
 }
+
+/// Compute firewall rules implied by an IP allowlist.
+///
+/// This looks up the source IP allowlist in the database, and generates
+/// firewall rules attached to each provided NIC that implements the allowlist.
+/// If the allowlist is "any", there are no restrictions and no rules are
+/// returned.
+///
+/// Otherwise, two sets of rules are returned:
+///
+/// - A low-priority rule that denies all traffic
+/// - A higher-priority rule that allows it from the specific IPs.
+async fn allowlist_to_firewall_rules(
+    nics: &[NetworkInterface],
+    allowed_ips: Vec<IpNet>,
+) -> Option<Vec<sled_agent_client::types::VpcFirewallRule>> {
+    if nics.is_empty() {
+        return None;
+    }
+
+    // Convert to host identifiers, how the sled-agent types expose these.
+    assert!(!allowed_ips.is_empty(), "Allowlist must not be empty here");
```

Comment:
This is probably fine, but I still don't love the `assert!`. I wonder if there's a way to structure `AllowedSourceIps::List` such that it statically guarantees the list isn't empty. I guess I kinda want `NonEmptyVec` like we have `NonZeroU32` and friends.

Thinking out loud - what could we do here if the list _is_ empty?

* deny all (which I think is what would happen if this assert wasn't here and we continued through the rest of the method - we'd add the `low_priority_deny` but then have an empty `filter_hosts` in the `allow` list, right?)
* panic (which is what would happen as written, given the assert)
* allow all

Allow all doesn't seem good! But panicking is maybe strictly worse than falling through to `deny all`? At least if we deny all Nexus's internal API still stays up, which seems good from an operator point of view, and a customer could even still get to Nexus's _external_ API via the techport-wicketd-proxy-over-the-underlay-network recovery path.

Reply:
Ok, a couple of good points here.

> I guess I kinda want `NonEmptyVec `

Yes! I could add a newtype around the `Vec<IpNet>` that enforces a non-empty list on the type level. I think that would need a bit of serde work using something like `serde(try_from = ...)`, but that's pretty easy.

> Allow all doesn't seem good! But panicking is maybe strictly worse than falling through to deny all? At least if we deny all Nexus's internal API still stays up, which seems good from an operator point of view, and a customer could even still get to Nexus's external API via the techport-wicketd-proxy-over-the-underlay-network recovery 

I had to convince myself on this one, but I think you're right. The key for me was that the underlay is not subject to these rules at all, so Nexus is still available, albeit in a frustrating way through the techport / wicketd proxy. Falling through to denying everything does seem better.

I think the non-empty-newtype obviates this right? Or maybe it masks the issue. If we don't get here because we fail that deserialization from the database, then won't apply _any_ rules, and we'll be back in the "allow all by default" case, which we don't want.

Reply:
Added the newtype in 381b1d2c0. If we do hit this error deserializing from the database, we return an error (instead of panicking). It's _sort of_ true that we'll fall through to deny all, but not necessarily. Consider this sequence:

- There's no allowlist when we first plumb the service rules, so the OPTE ports get the existing rules, which allow from any IP, but restricts ports / protocols.
- We somehow get an empty list in the database
- We try to plumb the rules again. We'll error out in this case, which just means we'll be stuck with the _existing_ rules, which are not "deny all". I don't think we have any better option, unless we wanted to try to handle this case and specifically forward a deny all rule. I don't think it's a great idea to try to force that work through at this point.